src/mistral_common/app/main.py

```diff
@@ -0,0 +1,336 @@
+
+
+def get_settings() -> Settings:
+    r"""Get the settings for the Mistral-common API."""
+    return Settings()
+
+
+@main_router.get("/")
+def redirect_to_docs() -> RedirectResponse:
+    r"""Redirect to the documentation."""
+    return RedirectResponse(url="docs")
+
+
+@main_router.get("/info")
+def get_info(settings: Annotated[Settings, Depends(get_settings)]) -> Dict[str, str]:
+    r"""Get the information about the Mistral-common API."""
+    return {"app_name": settings.app_name, "app_version": settings.app_version}
+
+
+def _tokenize_request(
+    request: Union[ChatCompletionRequest, OpenAIChatCompletionRequest],
+    settings: Annotated[Settings, Depends(get_settings)],
+) -> List[int]:
+    if isinstance(request, OpenAIChatCompletionRequest):
+        try:
+            request = ChatCompletionRequest.from_openai(**request.model_dump(exclude_none=True))
+        except (ValidationError, ValueError) as e:
+            raise HTTPException(status_code=400, detail=str(e))
+
+    if request.messages == []:
+        raise HTTPException(status_code=400, detail="Messages list cannot be empty.")
+
+    tokenized = settings.tokenizer.encode_chat_completion(request)
+    assert isinstance(tokenized, Tokenized), type(tokenized)
+    return tokenized.tokens
+
+
+def _tokenize_messages(
+    messages: list[ChatMessageType], settings: Annotated[Settings, Depends(get_settings)]
+) -> List[int]:
+    if len(messages) == 0:
+        raise HTTPException(status_code=400, detail="Messages list cannot be empty.")
+    request = ChatCompletionRequest(messages=messages)
+
+    return _tokenize_request(request, settings)
+
+
+@tokenize_router.post("/messages")
+def tokenize_messages(
+    settings: Annotated[Settings, Depends(get_settings)], messages: list[ChatMessageType] = Body(default_factory=list)
+) -> list[int]:
+    r"""Tokenize a list of messages."""
+    return _tokenize_messages(messages, settings=settings)
+
+
+@tokenize_router.post("/prompt")
+def tokenize_prompt(
+    settings: Annotated[Settings, Depends(get_settings)],
+    prompt: str = Body(default_factory=str),
+    add_special: bool = Body(default=True),
+) -> list[int]:
+    r"""Tokenize a prompt."""
+    if prompt == "":
+        raise HTTPException(status_code=400, detail="Prompt cannot be empty.")
+
+    return settings.tokenizer.instruct_tokenizer.tokenizer.encode(prompt, bos=add_special, eos=add_special)
+
+
+@tokenize_router.post("/request")
+def tokenize_request(
+    request: Union[ChatCompletionRequest, OpenAIChatCompletionRequest],
+    settings: Annotated[Settings, Depends(get_settings)],
+) -> list[int]:
+    r"""Tokenize a chat completion request."""
+    return _tokenize_request(request, settings=settings)
+
+
+@decode_router.post("/")
+def detokenize_tokens(
+    settings: Annotated[Settings, Depends(get_settings)],
+    tokens: list[int] = Body(default_factory=list),
+    as_message: bool = Body(default=False),
```

Comment:
let's only allow message to begin with -> smaller API surface is better in the beginning IMO